File: backend-go/service/race_service.go
package service

import (
	"context"

	"github.com/sargy/gridpulse/backend-go/model"
	"github.com/sargy/gridpulse/backend-go/repository"
)

type RaceService struct {
	repo *repository.RaceRepository
}

func NewRaceService(repo *repository.RaceRepository) *RaceService {
	return &RaceService{repo: repo}
}

func (s *RaceService) GetAllRaces(ctx context.Context) ([]model.Race, error) {
	return s.repo.FindAll(ctx)
}

func (s *RaceService) GetRacesBySeries(ctx context.Context, series string) ([]model.Race, error) {
	return s.repo.FindBySeries(ctx, series)
}

func (s *RaceService) GetRaceByID(ctx context.Context, id string) (*model.Race, error) {
	return s.repo.FindByID(ctx, id)
}

func (s *RaceService) Count(ctx context.Context) (int64, error) {
	return s.repo.Count(ctx)
}

func (s *RaceService) SaveAll(ctx context.Context, races []model.Race) error {
	return s.repo.InsertMany(ctx, races)
}

File: backend/src/main/java/com/gridpulse/api/service/RaceService.java
package com.gridpulse.api.service;

import com.gridpulse.api.model.Race;
import com.gridpulse.api.repository.RaceRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class RaceService {

    private final RaceRepository raceRepository;

    public RaceService(RaceRepository raceRepository) {
        this.raceRepository = raceRepository;
    }

    public List<Race> getAllRaces() {
        return raceRepository.findAll();
    }

    public List<Race> getRacesBySeries(String series) {
        return raceRepository.findBySeriesContaining(series);
    }

    public Optional<Race> getRaceById(String id) {
        return raceRepository.findById(id);
    }

    public long count() {
        return raceRepository.count();
    }

    public void saveAll(List<Race> races) {
        raceRepository.saveAll(races);
    }
}
